compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -3243,8 +3264,91 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
       }
     }
 
-    /** Gen JS code for a Labeled block from a pattern match, while trying
-     *  to optimize it away as an If chain.
+    /** Gen JS code for a Labeled block from a pattern match'es case, while
+     *  trying to optimize it away as a reversed If.
+     *
+     *  If there was no `return` to the label at all, simply avoid generating
+     *  the `Labeled` block alltogether.
+     *
+     *  If there was more than one `return`, do not optimize anything, as
+     *  nothing could be good enough for `genOptimizedMatchEndLabeled` to do
+     *  anything useful down the line.
+     *
+     *  If however there was a single `return`, we try and get rid of it by
+     *  identifying the following shape:
+     *
+     *  {{{
+     *  {
+     *    ...stats1
+     *    if (test)
+     *      return(nextCaseSym)
+     *    ...stats2
+     *  }
+     *  }}}
+     *
+     *  which we then rewrite as
+     *
+     *  {{{
+     *  {
+     *    ...stats1
+     *    if (!test) {
+     *      ...stats2
+     *    }
+     *  }
+     *  }}}
+     *
+     *  The above rewrite is important for `genOptimizedMatchEndLabeled` below
+     *  to be able to do its job, which in turn is important for the IR
+     *  optimizer to perform a better analysis.
+     *
+     *  This whole thing is only necessary in Scala 2.12.9+, with the new flat
+     *  patmat ASTs. In previous versions, `returnCount` is always 0 because
+     *  all jumps to case labels are already caught upstream by `genCaseBody()`
+     *  inside `genTranslatedMatch()`.
+     */
+    private def genOptimizedCaseLabeled(label: js.Ident,
+        translatedBody: js.Tree, returnCount: Int)(
+        implicit pos: Position): js.Tree = {
+
+      def default: js.Tree =
+        js.Labeled(label, jstpe.NoType, translatedBody)
+
+      if (returnCount == 0) {
+        translatedBody
+      } else if (returnCount > 1) {
+        default
+      } else {
+        translatedBody match {
+          case js.Block(stats) =>
+            val (stats1, testAndStats2) = stats.span {
+              case js.If(_, js.Return(js.Undefined(), Some(`label`)), js.Skip()) =>
+                false
+              case _ =>
+                true
+            }
+
+            testAndStats2 match {
+              case js.If(cond, _, _) :: stats2 =>
+                val notCond = cond match {
+                  case js.UnaryOp(js.UnaryOp.Boolean_!, notCond) =>
+                    notCond
+                  case _ =>
+                    js.UnaryOp(js.UnaryOp.Boolean_!, cond)
+                }
+                js.Block(stats1 :+ js.If(notCond, js.Block(stats2), js.Skip())(jstpe.NoType))
+
+              case _ =>
+                default
+            }
```

Comment:
I find reasoning about the correctness of this piece of code very difficult. Notably it's very hard to be sure that it is correct if there is a single `js.Return` but not in the shape we can optimize. So maybe add a `case Nil => default`, `case _ => fail(...)` to aid readability?